### ISSUE TITLE
High Availability guides should make distinction between single-site and multi-site deployments

### ISSUE THREAD

**Issue by tkyjovsk** [keycloak/keycloak#30095]
### Description

The "High availability" section of the [Keycloak guides overview page](https://www.keycloak.org/guides#high-availability) only contains items specific to multi-site deployments however Keycloak already supports HA in the single-site setup without any of this additional effort.

This might be misleading for some users, or at least not immediately clear (imo), as it may seem that these additional steps are needed to get HA.

To solve this I'd suggest:
A) to either rename the section to specifically "Multi-site high availability", or
B) to include an additional item at the start of the section for "Single site high availability" with a short explanation.

### Motivation

Clarification.

**Comment by shawkins** [comment]
Relates to #29388 - we are waiting for the operator to properly support scheduling features before including this guide.

**Comment by ahus1** [comment]
Related but not the same: Streched setups inside of a single region without an external Infinispan when you have transparent networking (either in EC2 setups, or in stretched OpenShift clusters): https://github.com/keycloak/keycloak/discussions/30147

**Comment by pschiffe** [comment]
Hello, given Keycloak 26 is out, `persistent-user-sessions` is enabled by default and external infinispan in a single-site setup is guarded with a feature flag (so it's not used) - is there at least some short summary somewhere how to make single site active-active HA work?
For example, with Keycloak operator, what do I have to do except defining `instances: 3`? (where DB & LB is taken care of)

**Comment by ahus1** [comment]
@pschiffe - there is nothing else to be done in addition to setting a value for instances greater than one, and you will have a single-site HA setup of Keycloak in Kubernetes.

In the docs, you find this today [in Support for Keycloak in a cluster](https://www.keycloak.org/server/configuration-production#_support_for_keycloak_in_a_cluster) and in the [introduction to multi-site](https://www.keycloak.org/high-availability/introduction) where we state "Keycloak supports deployments that consist of multiple Keycloak instances that connect to each other using its Infinispan caches; load balancers can distribute the load evenly across those instances. Those setups are intended for a transparent network on a single site."

The intent of this issue is to make it more explicit.

**Comment by ahus1** [comment]
Yes, that UDP and multicast based discovery is both a blessing and a curse: In some cloud environments it won't work, and in other environments you might discover Keycloak nodes of other clusters unintended. With KC 26.1, we're about to change this to use the Keycloak database for discovery by default, see https://www.keycloak.org/nightly/server/caching#_enable_distributed_caching

With Keycloak 26, sticky sessions are not required. In fact, they have never been required, but have been only a suggestion. [The docs say](https://www.keycloak.org/server/reverseproxy): "For performance purposes, it may be useful if load balancer forwards all requests related to particular browser session to the same Keycloak backend node."

See more of an explanation below. As I a summary I'd say it is quite cumbersome to setup sticky sessions, and might not be worth the added complexity. 

If you have a sticky session for a user, that user's information will be cached in Keycloak's local caches on subsequent requests, which might slightly increase the performance. If you do this IP based, this would can work, still it comes with a possibly unevenly balanced load when users are being corporate proxies or carrier grade NAT (like mobile users). There are ways to configure Keycloak to add specific cookies with node indicators as outlined in the docs. This is IMHO quite cumbersome. Some proxies also add their own cookies, and the docs describe how to leverage that. 

At the same time, using sticky sessions based on client IP addresses, a confidential client that connects directly to Keycloak and validates a token will by using sticky sessions always talk to the same Keycloak instance, thereby leading to an unevenly balanced load. And while a user's browser uses cookies to direct the traffic, a client never will, so it will usually end up on a different node compared to the user's browser, so the information won't be cached there.

**Comment by pschiffe** [comment]
Thank you for the great write up about the sticky sessions - I completely agree, I always try to avoid them if possible.

And great news about the new discovery protocol! Looking forward to the next release :blush:

**Comment by ahus1** [comment]
There is now an ongoing restructuring of the guides in #41544 targeted for 26.4, which should resolve this documentation gap.

In a nutshell: It describes "single-cluster" and "multi-cluster" deployments. This is part of our efforts to better test and support deployments in Kubernetes clusters that span multiple availability zones.
